### ISSUE TITLE
Why the loss function in ORPO compute nll loss respectively rather than use chosne_logps?

### ISSUE THREAD

**Issue by AIR-hl** [huggingface/trl#1676]
Why loss function in ORPO compute nll loss respectively rather than use chosne_logp? Is it better to directly use the existing `chosen_logps` to calculate?
```
def cross_entropy_loss(logits, labels):
    if not self.is_encoder_decoder:
        # Shift so that tokens < n predict n
        logits = logits[..., :-1, :].contiguous()
        labels = labels[..., 1:].contiguous()
    # Flatten the tokens
    loss_fct = nn.CrossEntropyLoss()
    logits = logits.view(-1, logits.shape[-1])
    labels = labels.view(-1)
    # Enable model parallelism
    labels = labels.to(logits.device)
    loss = loss_fct(logits, labels)
    return loss

if self.is_encoder_decoder:
    labels = concatenated_batch["concatenated_labels"].clone()
else:
    labels = concatenated_batch["concatenated_input_ids"].clone()
    attention_mask = concatenated_batch["concatenated_attention_mask"]
    labels = torch.where(attention_mask == 1, labels, self.label_pad_token_id)

chosen_nll_loss = cross_entropy_loss(all_logits[:len_chosen], labels[:len_chosen])
```

**Comment by mnoukhov** [comment]
Digging into it, looks like ORPO's nll loss does not mask the prompt tokens for the loss whereas the chosen_logps are masked. If you change the nll loss to just `-chosen_logps.sum()` this would be equivalent to doing a `CompletionOnly` SFT loss for ORPO. 

@jiwooya1000 your original code explicitly does this, is it a key component of ORPO or did you test without SFT on the prompt?

**Comment by jiwooya1000** [comment]
Hello @AIR-hl and @mnoukhov, thank you for the ping!

The original setting in ORPO uses the full NLL loss, including prompt loss.
While we did not have a chance to conduct an in-depth study on the impact of prompt loss in ORPO, using prompt loss had a higher MT-Bench when using the [capybara-7k](https://huggingface.co/datasets/argilla/distilabel-capybara-dpo-7k-binarized).

Some possible interpretations here could be:
1. Capybara-7k has a long prompt (i.e., long conversation history) and relatively short response, which makes learning from the given chat history helpful for the model
2. The weighting value ($\beta$) might need to be smaller without prompt loss. By excluding prompt loss, the magnitude of NLL loss will get smaller, leading to a stronger penalty for the rejected responses on the OR loss side with the same $\beta$. In that sense, smaller $\beta$, along with excluding prompt loss, could lead to a different scenario.

There are some papers that have different conclusions in using prompt loss for instruction-tuning, so it would be an interesting to study the impact of prompt loss in this setting too!

**Comment by AIR-hl** [comment]
> Hello @AIR-hl and @mnoukhov, thank you for the ping!
> 
> The original setting in ORPO uses the full NLL loss, including prompt loss. While we did not have a chance to conduct an in-depth study on the impact of prompt loss in ORPO, using prompt loss had a higher MT-Bench when using the [capybara-7k](https://huggingface.co/datasets/argilla/distilabel-capybara-dpo-7k-binarized).
> 
> Some possible interpretations here could be:
> 
> 1. Capybara-7k has a long prompt (i.e., long conversation history) and relatively short response, which makes learning from the given chat history helpful for the model
> 2. The weighting value (β) might need to be smaller without prompt loss. By excluding prompt loss, the magnitude of NLL loss will get smaller, leading to a stronger penalty for the rejected responses on the OR loss side with the same β. In that sense, smaller β, along with excluding prompt loss, could lead to a different scenario.
> 
> There are some papers that have different conclusions in using prompt loss for instruction-tuning, so it would be an interesting to study the impact of prompt loss in this setting too!

@jiwooya1000 Thanks for your explanation.：）
